src/xfieldhistory.h

```diff
@@ -0,0 +1,244 @@
+        }
+        assert(len == xfieldChanges.size());
+    }
+};
+
+class CBlockTreeDB;
+class UniValue;
+
+typedef std::map< TAPYRUS_XFIELDTYPES, XFieldChangeListWrapper > XFieldHistoryMapType;
+/*  This map is used instead of the list in federation params after v0.5.2
+ *
+ * XFieldHistoryMapType is a Global map to store all xfield changes.
+ *
+ * class CXFieldHistoryMap contains a static map of XFieldHistoryMapType.
+ * (All objects of this class read and write to the same map.) 
+ * This is the full list of xfield change in the active chain.
+ * When the isTemp flag is set in the constructor, all methods access the temp map.
+ *
+ * Note that this class is pure virtual. It is always accessed via CXFieldHistory or CTempXFieldHistory.
+ */
+class CXFieldHistoryMap{
+
+    const bool isTemp;
+
+protected:
+    static XFieldHistoryMapType xfieldHistory;
+    inline CXFieldHistoryMap(bool temp):isTemp(temp) { }
+
+public:
+    virtual ~CXFieldHistoryMap(){}
+    virtual XFieldHistoryMapType& getXFieldHistoryMap() const = 0;
+
+    template <typename T>
+    void GetLatest(TAPYRUS_XFIELDTYPES type, T & xfieldval) const {
+        auto& listofXfieldChanges = (isTemp ? this->getXFieldHistoryMap() : xfieldHistory).find(type)->second;
+        xfieldval = boost::get<T>(listofXfieldChanges.rbegin()->xfieldValue);
+    }
+
+    virtual XFieldChangeListWrapper& operator[](TAPYRUS_XFIELDTYPES type) const {
+        return xfieldHistory.find(type)->second;
+    }
+
+    virtual bool IsNew(TAPYRUS_XFIELDTYPES type, const XFieldChange& xFieldChange) const;
+    virtual void Add(TAPYRUS_XFIELDTYPES type, const XFieldChange& xFieldChange);
+    //void Remove(TAPYRUS_XFIELDTYPES type, const XFieldChange& xFieldChange);
+
+    virtual const XFieldChange& Get(TAPYRUS_XFIELDTYPES type, uint32_t height);
+    virtual const XFieldChange& Get(TAPYRUS_XFIELDTYPES type, uint256 blockHash);
+    int32_t GetReorgHeight();
+};
+
+/*
+ * class CXFieldHistory is an interface to access the global map of xfiled changes
+ * in the active chain which is guaranteed to be initialized.
+ * this defines ots own constructor and other utility methods that act on the data in CXFieldHistoryMap
+ */
+class CXFieldHistory : public CXFieldHistoryMap{
+
+public:
+    CXFieldHistory():CXFieldHistoryMap(false) {}
+    virtual ~CXFieldHistory(){}
+
+    //constructor to initialize the confirmed global map
+    inline explicit CXFieldHistory(const CBlock& genesis):CXFieldHistoryMap(false) {
+        xfieldHistory.emplace(TAPYRUS_XFIELDTYPES::AGGPUBKEY, XFieldChangeListWrapper(XFieldAggPubKey::BLOCKTREE_DB_KEY));
+        xfieldHistory.emplace(TAPYRUS_XFIELDTYPES::MAXBLOCKSIZE, XFieldChangeListWrapper(XFieldMaxBlockSize::BLOCKTREE_DB_KEY));
+
+        this->Add(TAPYRUS_XFIELDTYPES::AGGPUBKEY, XFieldChange(genesis.xfield.xfieldValue, 0, genesis.GetHash()));
+        this->Add(TAPYRUS_XFIELDTYPES::MAXBLOCKSIZE, XFieldChange(MAX_BLOCK_SIZE, 0, genesis.GetHash()));
```

Comment:
Right, genesis block does not contain the max block size in xfield. This is the same constant as before.  It is added to the list as the reference point from which the max block size is applicable. Code retrieving this data is generic when the initial values are all known. Otherwise there would more if conditions for each xfield type which is unnecessary.